mitiq/collector.py

```diff
@@ -0,0 +1,263 @@
+        for value, mult in zip(self._computed_results, collection.counts()):
+            results += [value] * mult
+        return results
+
+    def _call_executor(
+        self, to_run: Union[QPROGRAM, Sequence[QPROGRAM]]
+    ) -> None:
+        """Calls the executor on the input circuit(s) to run.
+
+        Args:
+            to_run: Circuit(s) to run.
+        """
+        result = self._executor(to_run)
+        self._calls_to_executor += 1
+
+        try:
+            result = list(result)
+            self._computed_results += result
+            self._executed_circuits += to_run
+        except TypeError:
+            self._computed_results.append(result)
+            self._executed_circuits.append(to_run)
+
+    @staticmethod
+    def is_batched_executor(executor: Callable) -> bool:
+        """Returns True if the input function is recognized as a "batched
+        executor", else False.
+
+        The executor is detected as "batched" if and only if it is annotated
+        with a return type that is one of the following:
+
+            * Iterable[float]
+            * List[float]
+            * Sequence[float]
+            * Tuple[float]
+            * numpy.ndarray
+
+        Batched executors can run several quantum programs in a single call.
+        See below.
+
+        Args:
+            executor: A "serial executor" (1) or a "batched executor" (2).
+
+                (1) A function which inputs a single `QPROGRAM` and outputs a
+                single expectation value as a float.
+                (2) A function which inputs a list of `QPROGRAM`s and outputs a
+                list of expectation values (one for each `QPROGRAM`).
+
+        Returns:
+            True if the executor is detected as batched, else False.
+        """
+        executor_annotation = inspect.getfullargspec(executor).annotations
+
+        return executor_annotation.get("return") in (
+            List[float],
+            Sequence[float],
+            Tuple[float],
+            Iterable[float],
+            np.ndarray,
+        )
+
+
+class CircuitCollection:
+    """A collection of circuits, nominally to execute on some backend."""
+
+    def __init__(self, circuits: Sequence[QPROGRAM]) -> None:
+        """Initializes a CircuitCollection.
+
+        Args:
+            circuits: Sequence of circuits to form the CircuitCollection.
+        """
+        self._raw_circuits = deepcopy(circuits)
+        self._cirq_circuits = [
+            convert_to_mitiq(circuit)[0] for circuit in self._raw_circuits
+        ]
+
+        # List of unique circuits.
+        self._unique: List[cirq.Circuit] = []
+
+        # Dictionary where each key refers to an index of self._raw_circuits
+        # that is the first occurrence of a unique circuit. The value of this
+        # key is the multiplicity/count (number of occurrences) of this circuit
+        # in self._raw_circuits. See CircuitCollection.unique_with_counts.
+        self._counts: Dict[int, int] = {}
```

Comment:
I would have thought that this would be of type `Dict[int, Circuit]`. Would is be simpler to replace `self._cirq_circuits` with a dict of type `Dict[int, Circuit]`?